[compiler-rt][CMake] Pass all flags to _Float16 try-compile

etcwilde opened the pull request:
The try-compile mechanism requires that `CMAKE_REQUIRED_FLAGS` is a space-separated string instead of a list of flags. The original code expanded `BUILTIN_FLAGS` into `CMAKE_REQUIRED_FLAGS` as a space-separated string and then would overwrite `CMAKE_REQUIRED_FLAGS` with `TARGET_${arch}_CFLAGS` prepended to the unexpanded
`BUILTIN_CFLAGS_${arch}`. This resulted in the first two arguments being passed into the try-compile invocation, but dropping the other arguments listed in `BUILTIN_CFLAGS_${arch}`.

This patch appends `${TARGET_${arch}_CFLAGS}` to the list, and then joins the elements with spaces, resulting in all arguments passed via `CMAKE_REQUIRED_FLAGS` being space-separated and being used in the try-compile when detecting Float16.

etcwilde committed:
[compiler-rt][CMake] Pass all flags to try-compile

The try-compile mechanism requires that `CMAKE_REQUIRED_FLAGS` is a
space-separated string instead of a list of flags. The original code
expanded BUILTIN_FLAGS into CMAKE_REQUIRED_FLAGS as a space-separated
string and then would overwrite CMAKE_REQUIRED_FLAGS with
TARGET_${arch}_CFLAGS prepended to the unexpanded
BUILTIN_CFLAGS_${arch}. This resulted in the first two arguments being
passed into the try-compile invocation, but dropping the other arguments
listed in BUILTIN_CFLAGS_${arch}.

This patch takes the contents of both TARGET_${arch}_CLFAGS and
BUILTIN_FLAGS_${arch}, combines them into a single list, and then
joins that list as a space-separated string in `CMAKE_REQUIRED_FLAGS`.

arichardson commented on compiler-rt/lib/builtins/CMakeLists.txt:
```diff
@@ -854,10 +854,12 @@ else ()
-      # BUILTIN_CFLAGS_${arch} is a CMake list, so we have to join it to create a valid command line.
-      list(JOIN BUILTIN_CFLAGS " " CMAKE_REQUIRED_FLAGS)
-      set(CMAKE_REQUIRED_FLAGS "${TARGET_${arch}_CFLAGS} ${BUILTIN_CFLAGS_${arch}}")
+      # CMAKE_REQUIRED_FLAGS must be a space separated string
+      # Join BUILTIN_CFLAGS_${arch} and TARGET_${arch}_CFLAGS as a
+      # space-separated string.
+      set(CMAKE_REQUIRED_FLAGS ${BUILTIN_CFLAGS_${arch}})
+      list(APPEND CMAKE_REQUIRED_FLAGS ${TARGET_${arch}_CFLAGS})
+      list(JOIN CMAKE_REQUIRED_FLAGS " " CMAKE_REQUIRED_FLAGS)
```
How about the following slightly simpler version?
```suggestion
      # CMAKE_REQUIRED_FLAGS must be a space separated string
      # Join BUILTIN_CFLAGS_${arch} and TARGET_${arch}_CFLAGS as a
      # space-separated string.
      set(_required_flags_list ${TARGET_${arch}_CFLAGS} ${BUILTIN_CFLAGS_${arch}})
      list(JOIN _required_flags_list " " CMAKE_REQUIRED_FLAGS)
```

arichardson reviewed (commented):
Unrelated to this change but I see `string(REPLACE ";" " " _TARGET_${arch}_CFLAGS "${TARGET_${arch}_CFLAGS}")` below, which I think could just be dropped and we use CMAKE_REQUIRED_FLASG instead?

etcwilde committed:
[compiler-rt][CMake] Simplify ARM VFP check

`TARGET_${arch}_CFLAGS` is expanded into `CMAKE_REQUIRED_FLAGS`, which
is implicitly added to the flags by `check_compile_definition`. We don't
need to re-expand the `TARGET_${arch}_CFLAGS` variable again.

etcwilde commented:
> Unrelated to this change but I see string(REPLACE ";" " " _TARGET_${arch}_CFLAGS "${TARGET_${arch}_CFLAGS}") below, which I think could just be dropped and we use CMAKE_REQUIRED_FLASG instead?

Even better, it looks like `check_compile_definition` implicitly adds `CMAKE_REQUIRED_FLAGS`:
https://github.com/llvm/llvm-project/blob/1edb6b0af1c47f88aa5c230a1a2b769eeb3c30aa/compiler-rt/cmake/Modules/CompilerRTUtils.cmake#L94-L103

I've dropped adding those arguments entirely and it's only adding `CMAKE_C_FLAGS`.

arichardson commented on compiler-rt/lib/builtins/CMakeLists.txt:
```diff
@@ -854,15 +854,15 @@ else ()
       if (${arch} MATCHES "^(arm|armhf|armv7|armv7s|armv7k|armv7m|armv7em|armv8m.main|armv8.1m.main)$")
-        string(REPLACE ";" " " _TARGET_${arch}_CFLAGS "${TARGET_${arch}_CFLAGS}")
-        check_compile_definition(__ARM_FP "${CMAKE_C_FLAGS} ${_TARGET_${arch}_CFLAGS}" COMPILER_RT_HAS_${arch}_VFP)
+        check_compile_definition(__ARM_FP "${CMAKE_C_FLAGS}" COMPILER_RT_HAS_${arch}_VFP)
```
I am not convinced that using cmake_c_flags here is correct but since that's an unrelated change it should be done separately.

petrhosek commented on compiler-rt/lib/builtins/CMakeLists.txt:
```diff
@@ -854,15 +854,15 @@ else ()
+      # CMAKE_REQUIRED_FLAGS must be a space separated string
+      # Join BUILTIN_CFLAGS_${arch} and TARGET_${arch}_CFLAGS as a
+      # space-separated string.
+      set(CMAKE_REQUIRED_FLAGS ${BUILTIN_CFLAGS_${arch}} ${TARGET_${arch}_CFLAGS})
+      list(JOIN CMAKE_REQUIRED_FLAGS " " CMAKE_REQUIRED_FLAGS)
```
This is going to overwrite `CMAKE_REQUIRED_FLAGS` set by parent directories. Instead, we should be concatenating the flags.
```suggestion
      set(BUILTIN_CFLAGS ${BUILTIN_CFLAGS_${arch}} ${TARGET_${arch}_CFLAGS})
      list(JOIN BUILTIN_CFLAGS " " BUILTIN_CFLAGS)
      set(CMAKE_REQUIRED_FLAGS "${CMAKE_REQUIRED_FLAGS} ${BUILTIN_CFLAGS}")
```

etcwilde committed:
[compiler-rt] include existing required flags

The original behavior also omitted any existing required flags. Patch
this to include them in the _Float16 support check.
